ksql-streams/src/main/java/io/confluent/ksql/execution/streams/StreamSourceBuilder.java

```diff
@@ -129,15 +149,22 @@ private static PhysicalSchema getPhysicalSchema(final AbstractStreamSource strea
       final AbstractStreamSource streamSource,
       final KsqlQueryBuilder queryBuilder,
       final Consumed<K, GenericRow> consumed,
-      final ValueMapperWithKey<K, GenericRow, GenericRow> mapper) {
-    return queryBuilder.getStreamsBuilder()
-        // 1. Create a KStream on the changelog topic.
-        .stream(streamSource.getTopicName(), consumed)
-        // 2. mapValues to add the ROWKEY column
-        .mapValues(mapper)
-        // 3. transformValues to add the ROWTIME column. transformValues is required to access the
-        //    streams ProcessorContext which has the timestamp for the record.
-        .transformValues(new AddTimestampColumn());
+      final Function<K, String> rowKeyGenerator
+  ) {
+    // for really old topologies, we must inject a dummy step to sure state store names,
+    // which use the node id in the store name, stay consistent:
+    final boolean legacy = queryBuilder.getKsqlConfig()
+        .getBoolean(KsqlConfig.KSQL_INJECT_LEGACY_MAP_VALUES_NODE);
+
+    KStream<K, GenericRow> stream = queryBuilder.getStreamsBuilder()
+        .stream(streamSource.getTopicName(), consumed);
+
+    if (legacy) {
+      stream = stream.mapValues(value -> value);
+    }
+
+    return stream
+        .transformValues(new AddKeyAndTimestampColumns<>(rowKeyGenerator));
```

Comment:
Guts of the PR's functional change is here.  Note that the `mapValues` call is only made for legacy queries and the actual work for adding ROWKEY is now in the transformer.